initData Correction : error in template, so initialize with empty values

mathiasv27 opened the pull request:
Hi,
I've made a little correction in iniData method. An error occured in 3.3.0.14, in settings.tpl : PHP Fatal error:  Uncaught TypeError: in_array(): Argument #2 ($haystack) must be of type array, null given
So the "additions" var is initialized with empty arrays if datas are not present.

NateWr commented:
Thanks @mathiasv27! It looks like your code editor has applied formatting across the file. Can you redo your PR so that only the bug fix is applied?

Also, can you remove the changes to the French `.po` file? All translations are applied through weblate. You can make changes for this plugin here: https://translate.pkp.sfu.ca/projects/plugins/texteditorextras/

mathiasv27 commented:
Ok, thanks ! I've reverted my commit and pushed only the bug fix, is that ok or do I have to cancel this request and redo a PR ?

NateWr commented on TextEditorExtrasSettingsForm.inc.php:
```diff
@@ -24,11 +24,31 @@ public function __construct($plugin) {
-		$this->setData('additions', $this->plugin->getSetting($contextId, 'additions'));
-		parent::initData();
-	}
+    public function initData()
```
In `stable-3_3_0` we are still using tabs not spaces. Can you convert to tabs so we're consistent with other code in this file?

NateWr commented on TextEditorExtrasSettingsForm.inc.php:
```diff
@@ -24,11 +24,31 @@ public function __construct($plugin) {
+                "editEmailTemplate" => array("body" => array()),
+                "titleAbstract" => array("abstract" => array()),
+                "announcement" => array("description" => array(), "descriptionShort" => array()),
+            );
```
Please use the short array syntax, `[]`, rather than the longer one, `array()` for all array declarations here.

NateWr commented on TextEditorExtrasSettingsForm.inc.php:
```diff
@@ -24,11 +24,31 @@ public function __construct($plugin) {
+                "announcement" => array("description" => array(), "descriptionShort" => array()),
+            );
+        if($this->plugin->getSetting($contextId, 'additions')){
+            foreach ($this->plugin->getSetting($contextId, 'additions') as $key => $value) {
```
Instead of looping over each item, I think you can use `array_merge`:

```php
$settings = $this->plugin->getSetting(...);
if (is_array($settings)) {
	$additions = array_merge($additions, $settings);
}
```

mathiasv27 replied:
I tried array_merge before doing that, but it doesn't work because arrays are empty

NateWr replied:
Can you try `array_merge_recursive`?

mathiasv27 replied:
Better with that, thx

NateWr reviewed (changes requested):
Thanks @mathiasv27! The revert commit is fine. I'll squash it all before I merge. A few comments here about code syntax, but otherwise looks good. :+1:

NateWr commented on TextEditorExtrasSettingsForm.inc.php:
```diff
@@ -24,9 +24,28 @@ public function __construct($plugin) {
 		$contextId = Application::get()->getRequest()->getContext()->getId();
-		$this->setData('additions', $this->plugin->getSetting($contextId, 'additions'));
+		$additions = [
+			"masthead" => ["description" => [] ],
```
Just one extra space at the end here: `[] ]` should be `[]]`.

NateWr commented:
That's all my comments addressed.  @mathiasv27 are you ready for me to merge this and port it to `main`?

mathiasv27 commented:
It's ok for me !

NateWr commented on TextEditorExtrasSettingsForm.inc.php:
```diff
@@ -24,9 +24,23 @@ public function __construct($plugin) {
+		];
+		$settings = $this->plugin->getSetting($contextId, 'additions');
+		if($settings){
+			$additions = array_merge_recursive($additions, $settings);
```
`$this->plugin->getSetting()` might not return an array, so can you check that it's an array before passing it to `array_merge_recursive()`?

```php
if (is_array($settings)) {
```

Thanks!

mathiasv27 commented:
Hi,
Is it possible to merge, or is something missing ?
Thanks a lot !

asmecher commented:
@mathiasv27, I think this is actually a duplicate of https://github.com/pkp/pkp-lib/issues/8064, which has already been fixed with additions to the template (rather than the .php code). What version of the plugin are you using?

mathiasv27 commented:
I'm using the last version on stable-3_3_0 : 1.0.0.2 . The additions to the template, such as `if is_array($additions.masthead.description)` didn't work, because if $additions or $additions.masthead is not initialized as an array, it throws an error...

asmecher commented:
@mathiasv27, ah! I think that's a mistake in using `is_array` instead of `isset`. Using `isset` instead would suppress the warning. Does that solve the problem in your case?

Your proposed solution of pre-initializing in PHP is also 100% valid, but with the additions already started in the template side, I'd favour continuing there. If you do prefer the PHP-side approach, please feel free to update the PR, but do also remove the template `is_array` calls in the process.

mathiasv27 commented:
Ok, that's better: I replaced is_array by isset and removed the is_array tests in the template. I prefer this approach indeed: sending a formatted and tested array of values to the template is, from my modest point of view, preferable to doing existence tests in a template :)
Thank you so much ! We have to present to users on our staging server shortly, do you think this can be merged soon?

asmecher commented:
@mathiasv27, agreed, this is a better solution. Thanks for working through it with us (and sorry for the delay)! I've merged it to `stable-3_3_0` and forward-ported it to `main` for the forthcoming 3.4.0 branch.

mathiasv27 commented:
> @mathiasv27, agreed, this is a better solution. Thanks for working through it with us (and sorry for the delay)! I've merged it to `stable-3_3_0` and forward-ported it to `main` for the forthcoming 3.4.0 branch.

Thanks a lot, Alec ! Don't worry about the delay, I understand ;)
